### DISCUSSION TITLE
Can open AudioStreamPreviewGenerator interface?

### DISCUSSION THREAD

**Discussion by colorworlds** [godotengine/godot-proposals#4505] [Editor]
I want to get an Audio Waveform for a music game.   
And I find this [module](https://github.com/v-pukman-gd/godot-audio-waveform-module/tree/master/audio_stream_preview_generator):  

This module have 4 core file:
```
audio_stream_preview.cpp
audio_stream_preview.h
audio_stream_preview_generator.cpp
audio_stream_preview_generator.h
```
But I found that the 4 core files in this module is Godot's own [code](https://github.com/godotengine/godot/blob/master/editor).

All the code in the module can be found in following 2 file:
```
audio_stream_preview.h
audio_stream_preview.cpp
```

So can Godot open this class directly?

**Comment by Calinou** [comment]
Audio visualization is a complex topic with no "one size fits all" answers. I think it's better for your project to write your own audio visualizer that actually suits your use cases. The Godot editor's waveform visualizer is intended to be used for editor preview purposes only, nothing else.

Also, since that code is in the `editor/` folder, it's only compiled in editor builds right now. To use it in a project, you'd need to also compile the code in question in export template builds, [making them larger in the process](https://docs.godotengine.org/en/latest/about/faq.html#why-does-godot-aim-to-keep-its-core-feature-set-small).

**Reply by KapFlagon** [reply]
>The Godot editor's waveform visualizer is intended to be used for editor preview purposes only, nothing else.

Would it be possible to expose the waveform visualizer or its API in some format that could allow re-use elsewhere in the editor? For example, audio focused plugins and add-ons could make use of the visualizer for tasks just within the editor, eliminating the need to include it in export builds.

**Comment by oeleo1** [comment]
I second @Calinou’s opinion. Adding the waveform plotting container node would be nice, but  it is too specific and too limited for becoming mainstream. With the advent of the audio processing tools at hand, you could happily write a much better, potentially multicolored waveform graphics representation of your audio samples. And I am talking from experience.
